File: app/src/main/java/org/stepic/droid/model/Section.java
package org.stepic.droid.model;

import android.os.Parcel;
import android.os.Parcelable;

import org.joda.time.DateTime;
import org.joda.time.DateTimeZone;
import org.joda.time.format.DateTimeFormat;
import org.joda.time.format.DateTimeFormatter;
import org.stepic.droid.base.MainApplication;
import org.stepic.droid.configuration.IConfig;
import org.stepic.droid.util.DateTimeHelper;

import java.io.Serializable;
import java.util.Locale;

import javax.inject.Inject;

public class Section implements Serializable, Parcelable {

    @Inject
    IConfig mConfig;

    private DateTimeFormatter mFormatForView;

    private long id;
    private long course; // course id
    private long[] units;
    private int position;
    private String progress;
    private String title;
    private String slug;
    private String begin_date;
    private String end_date;
    private String soft_deadline;
    private String hard_deadline;
    private String grading_policy;
    private String begin_date_source;
    private String end_date_source;
    private String soft_deadline_source;
    private String hard_deadline_source;
    private String grading_policy_source;
    private boolean is_active;
    private String create_date;
    private String update_date;
    private boolean is_cached;
    private boolean is_loading;

    public boolean is_loading() {
        return is_loading;
    }

    public void setIs_loading(boolean is_loading) {
        this.is_loading = is_loading;
    }

    public boolean is_cached() {
        return is_cached;
    }

    public void setIs_cached(boolean is_cached) {
        this.is_cached = is_cached;
    }

    private DateTime mBeginDateTime = null;
    private DateTime mSoftDeadline = null;
    private DateTime mHardDeadline = null;

    private String formatted_begin_date;
    private String formatted_soft_deadline;
    private String formatted_hard_deadline;

    public Section() {
        MainApplication.component(MainApplication.getAppContext()).inject(this);

        mFormatForView = DateTimeFormat
                .forPattern(mConfig.getDatePattern())
                .withZone(DateTimeZone.getDefault())
                .withLocale(Locale.getDefault());
    }

    public String getFormattedBeginDate() {
        if (formatted_begin_date != null) return formatted_begin_date;
        formatted_begin_date = DateTimeHelper.getPresentOfDate(begin_date, mFormatForView, mBeginDateTime);
        return formatted_begin_date;
    }

    public String getFormattedSoftDeadline() {
        if (formatted_soft_deadline != null) return formatted_soft_deadline;
        formatted_soft_deadline = DateTimeHelper.getPresentOfDate(soft_deadline, mFormatForView, mSoftDeadline);
        return formatted_soft_deadline;
    }

    public String getFormattedHardDeadline() {
        if (formatted_hard_deadline != null) return formatted_hard_deadline;
        formatted_hard_deadline = DateTimeHelper.getPresentOfDate(hard_deadline, mFormatForView, mHardDeadline);
        return formatted_hard_deadline;
    }


    public long getId() {
        return id;
    }

    public void setId(long id) {
        this.id = id;
    }

    public long getCourse() {
        return course;
    }

    public void setCourse(long course) {
        this.course = course;
    }

    public long[] getUnits() {
        return units;
    }

    public void setUnits(long[] units) {
        this.units = units;
    }

    public int getPosition() {
        return position;
    }

    public void setPosition(int position) {
        this.position = position;
    }

    public String getProgress() {
        return progress;
    }

    public void setProgress(String progress) {
        this.progress = progress;
    }

    public String getTitle() {
        return title;
    }

    public void setTitle(String title) {
        this.title = title;
    }

    public String getSlug() {
        return slug;
    }

    public void setSlug(String slug) {
        this.slug = slug;
    }

    public String getBegin_date() {
        return begin_date;
    }

    public void setBegin_date(String begin_date) {
        this.begin_date = begin_date;
    }

    public String getEnd_date() {
        return end_date;
    }

    public void setEnd_date(String end_date) {
        this.end_date = end_date;
    }

    public String getSoft_deadline() {
        return soft_deadline;
    }

    public void setSoft_deadline(String soft_deadline) {
        this.soft_deadline = soft_deadline;
    }

    public String getHard_deadline() {
        return hard_deadline;
    }

    public void setHard_deadline(String hard_deadline) {
        this.hard_deadline = hard_deadline;
    }

    public String getGrading_policy() {
        return grading_policy;
    }

    public void setGrading_policy(String grading_policy) {
        this.grading_policy = grading_policy;
    }

    public String getBegin_date_source() {
        return begin_date_source;
    }

    public void setBegin_date_source(String begin_date_source) {
        this.begin_date_source = begin_date_source;
    }

    public String getEnd_date_source() {
        return end_date_source;
    }

    public void setEnd_date_source(String end_date_source) {
        this.end_date_source = end_date_source;
    }

    public String getSoft_deadline_source() {
        return soft_deadline_source;
    }

    public void setSoft_deadline_source(String soft_deadline_source) {
        this.soft_deadline_source = soft_deadline_source;
    }

    public String getHard_deadline_source() {
        return hard_deadline_source;
    }

    public void setHard_deadline_source(String hard_deadline_source) {
        this.hard_deadline_source = hard_deadline_source;
    }

    public String getGrading_policy_source() {
        return grading_policy_source;
    }

    public void setGrading_policy_source(String grading_policy_source) {
        this.grading_policy_source = grading_policy_source;
    }

    public boolean is_active() {
        return is_active;
    }

    public void setIs_active(boolean is_active) {
        this.is_active = is_active;
    }

    public String getCreate_date() {
        return create_date;
    }

    public void setCreate_date(String create_date) {
        this.create_date = create_date;
    }

    public String getUpdate_date() {
        return update_date;
    }

    public void setUpdate_date(String update_date) {
        this.update_date = update_date;
    }


    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeLong(this.id);
        dest.writeLong(this.course);
        dest.writeLongArray(this.units);
        dest.writeInt(this.position);
        dest.writeString(this.progress);
        dest.writeString(this.title);
        dest.writeString(this.slug);
        dest.writeString(this.begin_date);
        dest.writeString(this.end_date);
        dest.writeString(this.soft_deadline);
        dest.writeString(this.hard_deadline);
        dest.writeString(this.grading_policy);
        dest.writeString(this.begin_date_source);
        dest.writeString(this.end_date_source);
        dest.writeString(this.soft_deadline_source);
        dest.writeString(this.hard_deadline_source);
        dest.writeString(this.grading_policy_source);
        dest.writeByte(is_active ? (byte) 1 : (byte) 0);
        dest.writeString(this.create_date);
        dest.writeString(this.update_date);
        dest.writeByte(is_cached ? (byte) 1 : (byte) 0);
        dest.writeByte(is_loading ? (byte) 1 : (byte) 0);
        dest.writeSerializable(this.mBeginDateTime);
        dest.writeSerializable(this.mSoftDeadline);
        dest.writeSerializable(this.mHardDeadline);
        dest.writeString(this.formatted_begin_date);
        dest.writeString(this.formatted_soft_deadline);
        dest.writeString(this.formatted_hard_deadline);
    }

    protected Section(Parcel in) {
        this.id = in.readLong();
        this.course = in.readLong();
        this.units = in.createLongArray();
        this.position = in.readInt();
        this.progress = in.readString();
        this.title = in.readString();
        this.slug = in.readString();
        this.begin_date = in.readString();
        this.end_date = in.readString();
        this.soft_deadline = in.readString();
        this.hard_deadline = in.readString();
        this.grading_policy = in.readString();
        this.begin_date_source = in.readString();
        this.end_date_source = in.readString();
        this.soft_deadline_source = in.readString();
        this.hard_deadline_source = in.readString();
        this.grading_policy_source = in.readString();
        this.is_active = in.readByte() != 0;
        this.create_date = in.readString();
        this.update_date = in.readString();
        this.is_cached = in.readByte() != 0;
        this.is_loading = in.readByte() != 0;
        this.mBeginDateTime = (DateTime) in.readSerializable();
        this.mSoftDeadline = (DateTime) in.readSerializable();
        this.mHardDeadline = (DateTime) in.readSerializable();
        this.formatted_begin_date = in.readString();
        this.formatted_soft_deadline = in.readString();
        this.formatted_hard_deadline = in.readString();
    }

    public static final Creator<Section> CREATOR = new Creator<Section>() {
        public Section createFromParcel(Parcel source) {
            return new Section(source);
        }

        public Section[] newArray(int size) {
            return new Section[size];
        }
    };
}

File: app/src/main/java/org/stepic/droid/model/Section.kt
package org.stepic.droid.model

import android.os.Parcel
import android.os.Parcelable
import com.google.gson.annotations.SerializedName
import org.stepic.droid.util.DateTimeHelper
import java.io.Serializable
import java.util.*

data class Section(
        var id: Long = 0,
        var course: Long = 0, // course id
        var units: LongArray = longArrayOf(),
        var position: Int = 0,
        var progress: String? = null,
        var title: String? = null,
        var slug: String? = null,
        @SerializedName("begin_date")
        var beginDate: String? = null,
        @SerializedName("end_date")
        var endDate: String? = null,
        @SerializedName("soft_deadline")
        var softDeadline: String? = null,
        @SerializedName("hard_deadline")
        var hardDeadline: String? = null,
        @SerializedName("grading_policy")
        var gradingPolicy: String? = null,
        @SerializedName("is_active")
        var isActive: Boolean = false,
        @SerializedName("create_date")
        var createDate: String? = null,
        @SerializedName("update_date")
        var updateDate: String? = null,
        var actions: org.stepik.android.model.actions.Actions? = null,
        @SerializedName("is_exam")
        var isExam: Boolean = false,
        @SerializedName("discounting_policy")
        var discountingPolicy: DiscountingPolicyType? = null,
        @SerializedName("is_requirement_satisfied")
        var isRequirementSatisfied: Boolean = true,
        @SerializedName("required_section")
        var requiredSection: Long = 0, //id of required section
        @SerializedName("required_percent")
        var requiredPercent: Int = 0) : Serializable, Parcelable {


    var isCached: Boolean = false
    var isLoading: Boolean = false

    val formattedBeginDate: String by lazy {
        DateTimeHelper.getPrintableOfIsoDate(beginDate, datePattern, TimeZone.getDefault())
    }

    val formattedSoftDeadline: String by lazy {
        DateTimeHelper.getPrintableOfIsoDate(softDeadline, datePattern, TimeZone.getDefault())
    }

    val formattedHardDeadline: String by lazy {
        DateTimeHelper.getPrintableOfIsoDate(hardDeadline, datePattern, TimeZone.getDefault())
    }


    override fun describeContents() = 0

    override fun writeToParcel(dest: Parcel, flags: Int) {
        dest.writeLong(id)
        dest.writeLong(course)
        dest.writeLongArray(units)
        dest.writeInt(position)
        dest.writeString(progress)
        dest.writeString(title)
        dest.writeString(slug)
        dest.writeString(beginDate)
        dest.writeString(endDate)
        dest.writeString(softDeadline)
        dest.writeString(hardDeadline)
        dest.writeString(gradingPolicy)
        dest.writeByte(if (isActive) 1.toByte() else 0.toByte())
        dest.writeString(createDate)
        dest.writeString(updateDate)
        dest.writeByte(if (isCached) 1.toByte() else 0.toByte())
        dest.writeByte(if (isLoading) 1.toByte() else 0.toByte())
        dest.writeParcelable(actions, flags)
        dest.writeInt(discountingPolicy?.ordinal ?: -1)
        dest.writeByte(if (isRequirementSatisfied) 1.toByte() else 0.toByte())
        dest.writeLong(requiredSection)
        dest.writeInt(requiredPercent)
    }

    protected constructor(input: Parcel) : this() {
        id = input.readLong()
        course = input.readLong()
        units = input.createLongArray()
        position = input.readInt()
        progress = input.readString()
        title = input.readString()
        slug = input.readString()
        beginDate = input.readString()
        endDate = input.readString()
        softDeadline = input.readString()
        hardDeadline = input.readString()
        gradingPolicy = input.readString()
        isActive = input.readByte().toInt() != 0
        createDate = input.readString()
        updateDate = input.readString()
        isCached = input.readByte().toInt() != 0
        isLoading = input.readByte().toInt() != 0
        actions = input.readParcelable(org.stepik.android.model.actions.Actions::class.java.classLoader)
        discountingPolicy = getDiscountingPolicyTypeByParcel(input)
        isRequirementSatisfied = input.readByte().toInt() != 0
        requiredSection = input.readLong()
        requiredPercent = input.readInt()
    }

    companion object {
        val datePattern = "dd MMMM yyyy HH:mm" //todo transfer to viewmodel/helper, keep section more plain

        @JvmField
        val CREATOR: Parcelable.Creator<Section> = object : Parcelable.Creator<Section> {
            override fun createFromParcel(input: Parcel): Section = Section(input)

            override fun newArray(size: Int): Array<Section?> = arrayOfNulls(size)
        }

        private fun getDiscountingPolicyTypeByParcel(input: Parcel): DiscountingPolicyType? {
            val temp = input.readInt()
            val localValues = DiscountingPolicyType.values()
            return if (temp >= 0 && temp < localValues.size) {
                localValues[temp]
            } else {
                null
            }
        }
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as Section

        if (id != other.id) return false
        if (course != other.course) return false
        if (!Arrays.equals(units, other.units)) return false
        if (position != other.position) return false
        if (progress != other.progress) return false
        if (title != other.title) return false
        if (slug != other.slug) return false
        if (beginDate != other.beginDate) return false
        if (endDate != other.endDate) return false
        if (softDeadline != other.softDeadline) return false
        if (hardDeadline != other.hardDeadline) return false
        if (gradingPolicy != other.gradingPolicy) return false
        if (isActive != other.isActive) return false
        if (createDate != other.createDate) return false
        if (updateDate != other.updateDate) return false
        if (actions != other.actions) return false
        if (isExam != other.isExam) return false
        if (discountingPolicy != other.discountingPolicy) return false
        if (isRequirementSatisfied != other.isRequirementSatisfied) return false
        if (requiredSection != other.requiredSection) return false
        if (requiredPercent != other.requiredPercent) return false
        if (isCached != other.isCached) return false
        if (isLoading != other.isLoading) return false

        return true
    }

    override fun hashCode(): Int {
        var result = id.hashCode()
        result = 31 * result + course.hashCode()
        result = 31 * result + Arrays.hashCode(units)
        result = 31 * result + position
        result = 31 * result + (progress?.hashCode() ?: 0)
        result = 31 * result + (title?.hashCode() ?: 0)
        result = 31 * result + (slug?.hashCode() ?: 0)
        result = 31 * result + (beginDate?.hashCode() ?: 0)
        result = 31 * result + (endDate?.hashCode() ?: 0)
        result = 31 * result + (softDeadline?.hashCode() ?: 0)
        result = 31 * result + (hardDeadline?.hashCode() ?: 0)
        result = 31 * result + (gradingPolicy?.hashCode() ?: 0)
        result = 31 * result + isActive.hashCode()
        result = 31 * result + (createDate?.hashCode() ?: 0)
        result = 31 * result + (updateDate?.hashCode() ?: 0)
        result = 31 * result + (actions?.hashCode() ?: 0)
        result = 31 * result + isExam.hashCode()
        result = 31 * result + (discountingPolicy?.hashCode() ?: 0)
        result = 31 * result + isRequirementSatisfied.hashCode()
        result = 31 * result + requiredSection.hashCode()
        result = 31 * result + requiredPercent
        result = 31 * result + isCached.hashCode()
        result = 31 * result + isLoading.hashCode()
        return result
    }

}
